vespalib/src/vespa/vespalib/util/spin_lock.h

```diff
@@ -0,0 +1,33 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <atomic>
+
+namespace vespalib {
+
+/**
+ * A spin-lock implementation that favors uncontended performance.
+ * Some measures are taken to reduce the impact of threads waiting to
+ * get the lock since this will not affect the fast-path of obtaining
+ * the lock immediately.
+ *
+ * This implementation satisfies the BasicLockable requirements,
+ * making it work with things like std::lock_guard.
+ **/
+class SpinLock {
+private:
+    std::atomic<bool> _lock;
+public:
+    SpinLock() noexcept : _lock(false) {
+        static_assert(std::atomic<bool>::is_always_lock_free);
+    }
+    void lock() noexcept {
+        while (__builtin_expect(_lock.exchange(true, std::memory_order_acquire), false)) {
```

Comment:
Are we on a new and cool™ enough version of GCC to be able to use the C++20 `[[likely]]` attribute?

Reply:
I am not sure, I thought C++20 was off the table for now, but I might be wrong... @baldersheim ?

Reply:
We are on 9 still, but could try to enable c++20 if it has support for it.

Reply:
I could not find it on the list: (maybe it has some cool name)
https://gcc.gnu.org/projects/cxx-status.html

But it does seem some C++20 features are available in gcc 9 (although experimental). We might want to wait until it is called 20 and not 2a...